### PR TITLE
Register AA Stream

### PR DESCRIPTION
**Pull request by Trevelopment**
This is like 90% there it works but the AA stream just doesn't make sound.  I think I have the wrong objectPath for the stream (callbacks.cpp - 278)
But this way I have it set up so you press :star: to play from USB and music button plays from your phone i'm sure we can just turn that into a toggle maybe even throw radio in there.
Oh and about all the weird repo stuff I reformatted this laptop so when I cloned this again I cloned it straight from master so that was set as the upstream and I forgot so I accidentally pushed to it when I meant to push to another branch and make a PR so I'm just makin some n00bstakes

### CODE REVIEW

File: mazda/callbacks.cpp [hunk 1]
```diff
@@ -261,6 +264,48 @@ std::string MazdaCommandServerCallbacks::GetLogPath() const
...
+	if(!aaStreamRegistered)
+	{
+		// so we dont accedentally try to register this twice
+		aaStreamRegistered = true;
+		try
+		{
+			// First open a new Stream
+			json sessArgs = {
+				{ "busName", "com.xsembedded.ServiceProvider" },
+				{ "objectPath", "/com/xse/service/AudioManagement/AudioApplication" },
```
Comment (Trevelopment) [inline mazda/callbacks.cpp]: @lmagder would you happen to know what the obectPath should be? I think its the final piece of this puzzle

**Comment by lmagder** [comment]
Hmm, I'm not sure why the stream registering doesn't work. It definitely seems like you are on the right track. Especially if you search for strings in the other CMU binaries that play audio. I noticed libjciamclient.so in jci/lib exports `AUDIO_AMSERVICE_registerAudioStream` and `AUDIO_AMSERVICE_openSession` functions that are used by the services, which we could call via dlopen, although we would need to guess at the arguments via the disassembly which would be a long and crash-filled task. They seem to just fire the D-Bus messages we are doing manually though, so not sure if that's really useful. I noticed the USB player passes "com.jci.usbm_am_client" and "/com/jci/usbm_am_client" when it creates the session so those seem to be the name of a DBus object the application is making previously to receive notifications from the audio manager.

Maybe there are some clues in /jci/am? there seem to be some .lua and .json files which define the stream types and how they interact.

Just to make sure though: Is the reason for adding a new stream to allow the existing USB player to work? I agree that registering a new stream instead of stealing the USB one is the right thing to do from a removing hacks point of view, but it may be really tricky to get working when hijacking the USB stream works-ish 🙃 . Although I guess if you solve the registering problem, we could register two streams, one Media type for AA_CH_AUD and one of type InfoUser for AA_CH_AU1 and the switching between the car audio and AA navigation and vice versa would all work perfectly.

**Comment by Trevelopment** [comment]
should I try this with com.jci.usbm_am_client?

**Comment by lmagder** [comment]
Yeah maybe, the notifications will go to the usb media service, possibly twice even, but maybe that's ok and has no effect? As far as i can tell the callback they are registering with DBus for this has the same signature as our AudioManagerClient::Notify so I think we are getting the same signals, just directly from the bus instead of via the CMU wrapper code, since they also compare against "is first param == "audioFocusChangeEvent"", etc.

**Comment by Trevelopment** [comment]
what about adding the stream to /jci/am/ControlLogic.lua?

**Comment by lmagder** [comment]
I'm not sure, but yeah it's possible they need to be in there. Would be kind of lame design on their part to be so hardcoded. I hope not, just from an install/uninstall/bricking POV though 😟 since the less system files we need to modify the better IMO.

I wonder if we could register a second "USB" stream in a different session id and it would work alongside the original one or if the audio manager would freak out? It seems the focus is based on session id, not the stream itself.

**Comment by Trevelopment** [comment]
You cant have 2 streams with the same name tried that but I am going to try com.xsembedded.service.xsesvc2dbus right now because that is the  busName used in ControlLogic.lua

**Comment by Trevelopment** [comment]
I have to be missing something that didnt work I even added the stream to ControlLogic.lua and it still didnt work.  What could I be doing wrong it registers the stream and everything just fine it just wont make any sound!!

**Comment by Trevelopment** [comment]
hey @lmagder just a question if you know this, what is the relationship between the streams used by the system and the channels used by AA.  I see this in `HUServer::hu_handle_ServiceDiscoveryRequest `
```
    audioChannel0->set_channel_id(AA_CH_AUD);
    {
      auto inner = audioChannel0->mutable_output_stream_channel();
      inner->set_type(HU::STREAM_TYPE_AUDIO);
      inner->set_audio_type(HU::AUDIO_TYPE_MEDIA);
      ...
   }
```
so does this mean the channels can only be separated by audio type? I didnt understand at first when you said to set the AA stream as type 'InfoUser' but now I see why you said that.  Do you think there is a way to set other streams of the same type as separate channels?

**Comment by lmagder** [comment]
@Trevelopment  the documentation on the AA side is sketchy since it's mostly reverse engineered but it seems like AA has at least two audio channels, AUDIO_TYPE_MEDIA which we are reporting as a 48kHz stereo 16 bit stream which is music and stuff and AUDIO_TYPE_SPEECH which is a 16kHz mono 16bit stream which is the navigation prompts, speech responses etc.

However we have assigned them to two protocol channels. It's unfortunate that they are both called "channels" since I think actually our mapping is arbitrary. AA only allows one "thing" per AA channel so we defined AA_CH_AUD and AA_CH_AU1 as "the two protocol channels we use for the two audio channels" but actually the AA_CH ids are arbitrary AFAIK.

We then open two ALSA sessions on the same ALSA device ("entertainmentUsb") with different bit rates and stereo settings to match the stream formats we told AA. Then internally the sound driver in the CMU OS mixes the two streams together (like if you were playing a video on a PC and playing music in media player at the same time, except that both sessions are in the same app). So actually both AA streams are playing whenever the car requests audio focus for either of them. However the AA software automatically mutes the media stream when the speech stream plays, so even though we are playing both the AUDIO_TYPE_MEDIA stream is silence.

> Do you think there is a way to set other streams of the same type as separate channels?

If there was a way to make two audio streams in the CMU audio manager that played from two different ALSA devices, we could open one session on each device instead of two on a single one. This would all the CMU audio manager to handle the audio focus more correctly since it can choose which of the two to play and AA can request focus individually like the protocol requires. Although I'm not sure what practical gain this would have other than it being nicer architecturally.

**Comment by lmagder** [comment]
Thanks. This pull request has been integrated manually.

**Comment by lmagder** [comment]
Wtf the comment above was automatically put there by my git client. Spooky. 

I merged your code to make the buttons work on the radio since the ungrab/grab stuff works great when I tried it. I also realized that my understanding of the AA audio focus I wrote above was wrong and it was making all the code super complex. I thought that the channel number in the message was the channel of the audio stream, but actually it's always AA_CH_CTR. So in reality you don't get AA messages per-channel the info is in the focus type enum itself, so AA only has a single focus state. However we were ignoring the enum values and turning it into a bool based on != LOST which was ignoring the transient/normal focus. So it allows you to tell the difference between media and prompts. So if there is no focus and there is a nav prompt or the mic it will as for transient focus, if you play music then it asks for normal focus. If it already has normal focus, a prompt sends no focus events. So really we just needed to keep that enum instead of tracking the channel list, which always was a single item.

If ifdefed out the call end stuff since it seemed like debugging code but the favorites button switching back to radio was useful. Before I had to close AA and then go into the "Entertainment" tab in the CMU to trigger the OS re-taking the focus. :+1:

**Comment by Trevelopment** [comment]
I Knew it!!  and I found the solution my friend

**Pull request merged**